pkgs/development/compilers/elm/packages/elm.nix

```diff
@@ -11,8 +11,8 @@ mkDerivation {
   version = "0.19.1";
   src = fetchgit {
     url = "https://github.com/elm/compiler";
-    sha256 = "1rdg3xp3js9xadclk3cdypkscm5wahgsfmm4ldcw3xswzhw6ri8w";
-    rev = "c9aefb6230f5e0bda03205ab0499f6e4af924495";
+    sha256 = "1h9jhwlv1pqqna5s09vd72arwhhjn0dlhv0w9xx5771x0xryxxg8";
+    rev = "2f6dd29258e880dbb7effd57a829a0470d8da48b";
```

Comment:
I’m a bit late to the game (and I don’t even use Nix, I just ended up here 😄), but this is _not_ the commit that the any official Elm binaries was built from.

All binaries except the new macOS arm64 binary are built from https://github.com/elm/compiler/commit/c9aefb6230f5e0bda03205ab0499f6e4af924495 (the removed commit in this diff), but new equivalent ones could be built the same way that the macOS arm64 one was built, so that doesn’t matter.

With that out of the way, how was the macOS arm64 binary built then? It’s built from the good old https://github.com/elm/compiler/commit/c9aefb6230f5e0bda03205ab0499f6e4af924495 commit, plus just this commit: https://github.com/elm/compiler/commit/0421dfbe48e53d880a401e201890eac0b3de5f06

How can I know that? Because I was involved in making unofficial macOS arm64 binaries. In early attempts, we created it from latest `master`. Then we ran into two edge cases:

- An elm-watch snapshot test containing an Elm error message started failing, because the snapshot was made with the x86 binary but our new arm64 binary had this fix in it: https://github.com/elm/compiler/commit/b5905470bea41bead435edd9f1b11b33ad8625c6. Having a typo-fixed error message is of course good, but it shows that the binaries aren’t equivalent and that the official arm64 wasn’t built with all commits on master because it also has the typo.
- This one is more important: Some change on master since the 0.19.1 has ever so slightly changed in which order some declarations in the compiled JS are emitted. This caused some hard-to-debug CI failures for me, where a hash matched on my macOS arm64 computer, but not on x86 GitHub Actions.

Examples:

```diff
         }, function(_v4) {
-            var limit = _v4.limit;
-            var offset = _v4.offset;
             var unit = _v4.unit;
+            var offset = _v4.offset;
+            var limit = _v4.limit;
             return _Utils_Tuple3(limit, offset, unit);
```

And:

```diff
         var $author$project$Types$TimeOfDay$toString = function(_v0) {
-            var hour = _v0.a.hour;
             var minute = _v0.a.minute;
+            var hour = _v0.a.hour;
             return A3(
```

So if you want to have Nix build binaries equivalent to the official ones, you need to take the old 0.19.1 release commit + the GHC 9 patch. That’s how Homebrew does it, btw: https://github.com/Homebrew/homebrew-core/pull/152453

Reply:
thanks @lydell I will look further into this. For nixpkgs we really want to switch to ghc9 because for old GHC8 we never really had a compiler and correct dependency versions to build with so we always compiled in jail-braked version (removing version constraints) for very long time which now case problems. Nixpkgs distribution was different from the official binaries for a very long time because of this jail break but perhaps it would make sense to get it as close as possible to official one.

Reply:
It’s up to you if you want it to be as close as possible to the official one, now you know about it at least. :+1: